catalog/core/catalog-core-commons/src/main/java/ddf/util/XPathHelper.java

```diff
@@ -156,13 +159,17 @@ public static String print(Node n, String encoding) {
       StringWriter stringOut = new StringWriter();
 
       DOMImplementationLS domImpl = (DOMImplementationLS) document.getImplementation();
-      LSSerializer serializer = domImpl.createLSSerializer();
-      LSOutput lsOut = domImpl.createLSOutput();
-      lsOut.setEncoding(encoding);
-      lsOut.setCharacterStream(stringOut);
-
-      serializer.write(n, lsOut);
-
+      LSSerializer lsSerializer = domImpl.createLSSerializer();
+      lsSerializer.getDomConfig().setParameter(XML_DECLARATION, false);
+      LSOutput lsout = domImpl.createLSOutput();
+      lsout.setEncoding(encoding);
+      lsout.setCharacterStream(stringOut);
+
+      NodeList childNodes = n.getChildNodes();
+      // Loop through nodelist in order to have it process xml attributes correctly DDF-4382
+      for (int i = 0; i < childNodes.getLength(); i++) {
```

Comment:
❓ Just want to make sure I'm understanding this correctly, previously we weren't looking at these childnodes but now we are?

Reply:
correct. Before, whenever a node that was printed that had XML attributes on it it would not print them and return empty string

Reply:
Ah okay that makes a lot of sense. I could see how this would get buggy wherever XPathHelper is used!

Reply:
Clarification for future. I have changed this for loop logic. When a node is of type NODE.ATTRIBUTE_NODE, its child element is a NODE.TEXT_NODE version of itself. Therefore, instead of serializing an ATTRIBUTE_NODE which cannot be done through the serializer, we get it as a TEXT_NODE and serialize using that that.